pkg/proctree/tree.go

```diff
@@ -0,0 +1,297 @@
+package proctree
+
+import (
+	"fmt"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/aquasecurity/tracee/pkg/logger"
+	"github.com/aquasecurity/tracee/pkg/utils/types"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+// nodeUniqueId is a set of information which should be unique enough to identify a node from
+// another at all cases.
+type nodeUniqueId struct {
+	id         int
+	uniqueTime int
+}
+
+// gcNode is a node which support being cleaned from the process tree.
+// All the nodes followed by the process tree has to implement this interface.
+type gcNode interface {
+	// GetUniqueId generates the unique ID to identify the node with in the tree.
+	GetUniqueId() nodeUniqueId
+	// DisconnectNode remove all references from the node to other nodes and vice versa.
+	// There should be no connections to and from the node after this is called.
+	DisconnectNode()
+}
+
+type ProcessTreeConfig struct {
+	// Max amount of processes nodes to allow in memory.
+	// Too small value here might result missing information in the tree and inconsistency.
+	// Default recommended value is 32768, as this is the max amount of PIDs in the system.
+	// You might even want more than 32768 to allow relevant information of dead processes to be
+	// available after their exit.
+	MaxProcesses int
+	// Max amount of threads nodes to allow in memory.
+	// Too small value here might result missing information in the tree and inconsistency.
+	// Default recommended value is 32768, as this is the max amount of TIDs in the system.
+	// You might even want more than 32768 to allow relevant information of dead threads to be
+	// available after their exit.
+	MaxThreads int
+	// Max size of cache for processes removing after exit - translate to the delay between process
+	// exit to its removal from the tree (if it is not interesting after its death).
+	// Too small value here might result processes missing from tree after their exit, before
+	// users were able to query the process information because of the delay between the feeding
+	// of the tree and the consumption of its information.
+	MaxCacheDelete int
+}
+
+// ProcessTree is a struct which follow the state of processes during runtime of a system.
+// The process tree is updated through Tracee's events, and is designed to overcome problems that
+// may arise because of events consumption (like handling lost events).
+// For more information on the logic, please go to the package documentation.
```

Comment:
Where is the package documentation ? I believe the logic should be explained interleaved with the code, so its easier to understand for first readers, IMO.

Reply:
The package documentation is written currently in the `export.go` file.
I don't want to have duplicate of documentation, because it always leads to one of them is not up-to-date with the other.